src/chunk.c

```diff
@@ -1536,6 +1537,160 @@ ts_chunk_get_by_relid(Oid relid, int16 num_constraints, bool fail_if_not_found)
 	return chunk_get_by_name(schema, table, num_constraints, fail_if_not_found);
 }
 
+static Oid chunk_for_tuple_oid = InvalidOid;
+static inline Oid
+get_chunk_for_tuple_oid()
+{
+	if (chunk_for_tuple_oid == InvalidOid)
+	{
+		Oid params[] = { INT4OID, ANYELEMENTOID };
+		chunk_for_tuple_oid = get_function_oid("chunk_for_tuple", INTERNAL_SCHEMA_NAME, 2, params);
+		Assert(chunk_for_tuple_oid != InvalidOid);
+	}
+	return chunk_for_tuple_oid;
+}
+
+bool
+ts_is_chunk_for_tuple_target_entry(TargetEntry *te)
+{
+	FuncExpr *fe;
+
+	if (!IsA(te->expr, FuncExpr))
+		return false;
+
+	fe = (FuncExpr *) te->expr;
+	if (fe->funcresulttype != INT4OID || list_length(fe->args) != 2)
+		return false;
+
+	if (fe->funcid != get_chunk_for_tuple_oid())
+		return false;
+
+	return true;
+}
+
+/* Process all the target entries to find the optimization.
+ *  Return whether or not the optimization was performed */
+bool static chunk_for_tuple_opt_check_target_entries(PlannedStmt *stmt, List *target_list)
+{
+	ListCell *lc;
+	foreach (lc, target_list)
+	{
+		TargetEntry *te = lfirst(lc);
+		FuncExpr *fe;
+		ConvertRowtypeExpr *crte;
+		Var *var;
+		RangeTblEntry *rte;
+		Chunk *chunk;
+
+		Assert(IsA((Node *) te, TargetEntry));
+
+		if (!ts_is_chunk_for_tuple_target_entry(te))
+			continue;
+
+		fe = (FuncExpr *) te->expr;
+		if (!IsA(lsecond(fe->args), ConvertRowtypeExpr) || !IsA(linitial(fe->args), Const))
+			continue;
+
+		crte = (ConvertRowtypeExpr *) lsecond(fe->args);
+
+		if (!IsA(crte->arg, Var))
+			continue;
+
+		var = (Var *) crte->arg;
+		rte = rt_fetch(var->varno, stmt->rtable);
+		chunk = ts_chunk_get_by_relid(rte->relid, 0, false);
+		if (chunk == NULL)
+			continue;
+
+		/* sanity check */
+		if (chunk->fd.hypertable_id != DatumGetInt32(((Const *) linitial(fe->args))->constvalue))
+			continue;
+
+		te->expr = (Expr *)
+			makeConst(INT4OID, -1, InvalidOid, 4, Int32GetDatum(chunk->fd.id), false, true);
+
+		/* we only expect one such entry so it's ok to return here */
+		return true;
+	}
+	return false;
+}
+
+/* Recurse throughout the plan to do this optimization */
```

Comment:
you are not recursing properly here and your logic will only work for toplevel elements, this needs more tests where toplevel is not append, i think mergeappend will not work with the current implementation